### ISSUE TITLE
Opencart 2.3's default behavior is to iset all links in admin to use HTTPS, 

### ISSUE THREAD

**Issue by lucasjkr** [copona/copona#73]
I know it's obviously best practice to only access the admin area through a secure connection (and, indeed, every site), but I think the reality is that many people (or, at least me!) develop on local machines that don't have SSL.

I didn't realize how much OC had changed in 2.3, where the setting to change whether or not to use SSL was removed from the following files:

/system/config/admin.php
/system/config/catalog.php

Where previously toggling a setting in admin.php could let you access the admin without SSL, that's been removed. Instead, the OC developers have hardcoded SSL into all the link functions in the admin section, without regard for whether the site owner has actually specified to use SSL or not (in 2.3, jf there is a value for HTTPS_SERVER in config.php, it uses that, rather than examining what's in the database)

I began the process in my fork of Copona - referring the the configuration setting in the database, and removing the hard-coded requirement to use SSL. I didn't realize what a can of worms that was going to open - according to PHPStorm, there are 2501 references that will need to be changed to support this (not that difficult to do with find and replace, but I'm sure it would be an alarming looking pull request).

Therefore, I wanted to ask what you all thought of it before continuing along. You can refer to the recent commits in my fork to see what I am referring to.

**Comment by arnisjuraga** [comment]
I agree, that SSL and no-SSL usage is confusing in Opencart,  specificly: ` $this->url->link('common/dashboard', '', true);` - is not a problem.
But this job you are going to do, is not needed.

In OC1 and OC2 minor version, there was a problem if you don't specify third paremeter, like:

` $this->url->link('common/dashboard', '');` - then there would be NO WAY to generate HTTPS in links, because of how `url->link` method works.

It could be the reaseon, why in the latest OC versions, they have added third paremeter as "true" by default. Instead of fixing or changing the url->link and settings itself, to be intuitive, not like 

https://github.com/opencart/opencart/issues/4102
https://github.com/opencart/opencart/issues/4158
https://github.com/opencart/opencart/commit/057f12e63aba20e3259234db322ad128afc9b91c
and so on.

But this doesn't mean that all links generate as HTTPS. 
`url->link` method, even though 'ssl' or 'true' has been passed, still  checks configuration and generates HTTPS only if:
- you have configured that in System -> Settings
- have defined HTTP_SERVER
and I think there was something else too.

So, I think, there is no need to rewrite it. Leave it as is.

The thing we would need to do - create simple configuration settings and flow so that SSL and no-SSL can be configured very simple.

At the moment - there are **too much different places and ways** how does SSL must be configured;
- settings in admin System > Settings > Server
- 'true' or 'SSL' parameter in url->link
- `defined` constants HTTP_SERVER and HTTPS_SERVER
- fallbacks in /system/config ... 
(grr)

So, there should be straightforward way to do it:

- set it to yes/no for admin/catalog
- additionally - have different URL for admin, catalog (if really needed)
- have a possibility to add additional domains/aliases for CDN functionality and configure no-https for them e.g.

**Comment by lucasjkr** [comment]
Let me clarify. 

In Opencart 2.3, it seems to ignore the configuration setting altogether. Rather, it goes by the value given to the HTTPS_SERVER constant in config.php.

You end up using HTTPS if you set the HTTPS_SERVER value to https:// rather than http:// (as it gets set by default in the installation process).

Further, all of the admin area passes "true" to the link builder (forcing it to use HTTPS) rather than letting the user configure it at all.

My goal was to poll the actual settings that the user selected and use those, pass the appropriate value to url->link based on what was selected in System > Settings > Server

We could add more granularity in the settings - rather than having one option for security, have "catalog uses SSL" and "admin uses SSL", and have the url->link work based on that.

In which case we should (I think) stop passing true/false to the link builder at all.

If we have link builder fix it, we should then clean up all the calls to it so we're not passing unused data into the function.

...

New idea:

Change settings to:

Use SSL for Catalog (y/n)
Use SSL for Admin (y/n)

And have url->link use to determine whether to build an HTTP or HTTPS link.

Once that's done, we don't need the true/false being fed in to the function on a link by link basis, and should eventually remove all references to it.

This is one of those instances where, while I enjoy opencart and it's easy to use, I feel like they throw "features" in without regard to how the system they built actually works...

Lets come up with an idea we all agree on, and that'll be something for me to work on.

**Comment by arnisjuraga** [comment]
Yes, I like an idea.
Althougt, I'am changing my mind - no need to over complicate :)
I mean, there is no need for SSL settings on/of divided admin/catalog. Because all modern browsers are (will be) forcing to use SSL for any web pages. [Google Will Soon Shame All Websites That Are Unencrypted](https://motherboard.vice.com/en_us/article/google-will-soon-shame-all-websites-that-are-unencrypted-chrome-https )
 and already Chrome in latest versions gives warnings (not yet in the aggressive way) for unencrypted password forms. Nowadays, SSL certificates can be issued for free using e.g. [Let's Encrypt!](https://letsencrypt.org/)

**So my proposal:**
- remove unneeded 'ssl' or 'true' in any files for url->link generation. All corresponding files can be easily searched and even replaced with regex (e.g. using Notepad++).
- make settings permanent but possible to override with some config.php (for foolproof solution - if you switch SSL on hostings which does not support HTTPS or does not redirect correctly and so on). I mean - first check something like:
~~~php
// in system/library/url.php create private $secure
private $secure; 
[..]
// in constructor, set the value
$this->secure= $this->config->get('config_secure'); 
//IF constant  'NOSSL' is defined, override and use no-ssl
$this->secure= defined('NOSSL') ? false : $this->secure;
//[..]
// in link() method, remove unneded parameters, and befor generating url, 
// check, if($this->secure)
...
~~~
- make simple PHP header() redirect from http:// to https:// in `startup.php` to avoid using `.htaccess` 
 - this must be done _after_ url class has been instantinated.
- installation script should also use https:// if hosting has it.

**Comment by lucasjkr** [comment]
I was thinking about simplification issues as well, and came up with this.

How about:

A) remove the 'config_secure' setting from Admin altogether

B) responding to the request using the same protocol that was used to generate it.

C) adding a new setting in config PHP called "FORCE_SSL" with a default value of "true" when installing.

D) using the header redirect you suggest in startup.php, so that if FORCE_SSL is true, they get the HTTPS version back even if they accessed it over HTTP, and therefore future requests would be made via HTTPS.

E) So first, we'll just set the URL class to ignore the "true" and "falses" that get passed into it, and once we're satisfied with the results, then we can clean up all the references to it.

I agree that deployed installations shouldn't be using insecure connections, but I don't set up HTTPS on my dev box (it's just virtual box with bridged networking, so therefore can't get Letsencrypt either). A friend that also does OpenCart development has the same headaches when he develops locally as well. So my thought is this is for development purposes.

**Comment by lucasjkr** [comment]
> installation script should also use https:// if hosting has it.

I was thinking about this one. 

Could be easy or difficult - some hosts serve HTTPS from the same directory as HTTP, others actually serve from a different directory.

The easy way would be, if the install is accessed over https, we can feel safe to assume that SSL is available.

The difficult way would involve checking to see if the page was also available via HTTPS rather than blindly redirecting them to it. Perhaps add a unique string to a meta tag in install, and when the page is first loaded, we do a curl request to https://SERVER_NAME, and scan the resulting file looking for our string. If we receive it, then we redirect to https, otherwise assume it's not available.

This would only be during the setup//install phase - its not like we'd be generating a curl request to ourselves with every page load.

Let me know your thoughts. I have a bit of free time tomorrow and would be happy to plunk away at this for a bit, so long as I know I'm taking an approach that we all agree on.

**Comment by arnisjuraga** [comment]
Agreed, can you make pull request?

About installation process - there should be the same workflow - IF https is available, then use https, 
otherwise - use http.

In fact - there could be just simple installation button in Step 1 - Install using HTTPS, Install Using HTTP.

As you said - there is no need to force use https always and everyewhere. Copona must be simple deployed and used with or without https (either live or dev box).
This just must be really easy configurable.

**Comment by lucasjkr** [comment]
I will redo what I've done, have url->Link deliver links in the protocol that the page was requested in, and then make a pull request.

I'd also like to do one that's does more extensive PHP version checking, rather than just changing the string on the screen.

**Comment by arnisjuraga** [comment]
Nice one.
@lucasjkr - congrats on first commits! Thank you!

**Comment by ADDCreative** [comment]
The problems with the HTTPS configuration lies with the usage of config variables site_ssl, config_ssl and config_secure.

The problem starts in the config files in system/config/ https://github.com/copona/copona/search?utf8=%E2%9C%93&q=site_ssl

As you can see it's not clear what its function is. Is it a flag of the HTTPS address of the server? It seems to be used as both.

The next problem is the new Url object is created in the framework using site_ssl regardless. This is what makes the admin always use HTTPS_SERVER regardless of the 'Use SSL' setting in the admin panel, as a new URL object is not created in the admin startup.php.

Next we come to the Url object in the catalog. https://github.com/copona/copona/blob/master/catalog/controller/startup/startup.php#L159
Again this is set using config_ssl regardless of the 'Use SSL' setting in the admin panel.

I believe the line should be something like below, test config_secure (which is set by the 'Use SSL' setting in the admin panel).
`$this->registry->set('url', new Url($this->config->get('config_url'), $this->config->get('config_secure') : $this->config->get('config_ssl') : ''));`

A similar line should be added to admin/startup/startup.php, but using site_base and site_ssl.

The problem is not with the Url Link method, but with the startup configuration.

**Comment by lucasjkr** [comment]
I'm going to disagree. There are some settings that shouldn't be configurable by a user, even if they have "admin" privileges to the store. Database name, db host, db username, password, and whether or not to use SSL. 

In many cases the person with admin privs, this isn't the person who is responsible for the site, and they might change a setting without understanding the ramifications. It's like setting the database name database user in the web interface - the store admin shouldn't even be given that option.

If the developer needs to turn off HTTPS, it should be done through modifying the config file, something that normal users wouldn't ever think to do (or hopefully, be able to do), but which is obvious (and will be documented) for developers and anyone else responsible for maintaining an installation.

I think the easiest solution which addresses your concern and Arnis is just by adding the force_ssl global; ordinary stores would have it set to true by default, and we would include instruction that stores in "development" mode (without SSL certs) or behind load balancers should need to change that setting to false, if they are part of the limited number of edge cases that don't need it.

**Comment by ADDCreative** [comment]
@lucasjkr I agree that the best place for selecting to use HTTPS may by be in the config files. I was just trying to point out the bug is with the startup code not the Url Link method. By all means change the startup code and remove the 'Use SSL' setting from the admin panel. There is nothing wrong with doing that,

There is a problem with setting the link protocol based on the request. Doing so is not secure and should be avoided.

The problem is not with the link builder link method, but how the link builder is configured when created.

I can explain more if required.

**Comment by lucasjkr** [comment]
My point is I would like a universal, configurable method of turning SSL on and off.

I understand your concern about how I implemented it in the link builder, again, in "my vision" (Sorry for not sharing, just my first contribution without much conversation), ordinary sites would simply only serve pages via HTTPS, and therefore the link builder becomes a non-issue.

I don't agree with OC's 2.3 method of passing data into the link builder to determine whether or not a link should be secure or not, that's pretty arbitrary and makes things a pain compared to how it worked in 2.2

If rely on the config.php file for whether or not to create secure links, that means that making the site secure becomes a one or the other proposition. Whereas this way, we have more flexibility. Again, the purpose for this was to allow the site admin to be able to configure the site prior to obtaining an SSL certificate if necessary.

if you have an idea to improve it, I would of course love to hear your thoughts on it, I'll keep an open mind as I hope you will too!

**Comment by ADDCreative** [comment]
I understand what you're saying and are trying to do. It just the way it's been implemented in https://github.com/copona/copona/pull/74 is wrong and creates an insecurity.

Now if an attacker can get an admin or customer to visit a login page via a non secure URL (for example http://example.com/admin/index.php?route=common/login, note the http not https) the form with the username and password will be summited over a nonsecure link (as the request was not secure, the links generated were not secure). Therefore giving an opportunity for the details to be stolen.

As you have seen changing the HTTPS configuration is complicated and has many pitfalls. Doing it wrong can cause security problems.

A few notes to keep in mind.

Both OpenCart 2.2 and 2.3 were broken in selecting HTTPS. So I would not go by how either of those versions worked. Best to work out a new method and correctly implement it.

You also have to take in to account how a multi stores setup works. Here the HTTP and HTTPS URLS are entered in the admin panel and the the HTTPS is optional. This will probably need changing. Would you have one option to switch HTTPS on and off for all stores or as it is at the moment where each store is configured individually?

Any page that should be secure (all of the admin and at least the account and checkout of the catalog) should really check that it was indeed access over HTTPS (somthing we added to CE). I think @arnisjuraga also mentioned this.

**Comment by lucasjkr** [comment]
Sorry, just waking up to continue this discussion.

> I understand what you're saying and are trying to do. It just the way it's been implemented in #74 is wrong and creates an insecurity... Now if an attacker can get an admin or customer to visit a login page via a non secure URL (for example http://example.com/admin/index.php?route=common/login

In OC's default setting as currently implemented, if the site is set up initially by accessing it over http://, that's the value that gets written to both the HTTP_SERVER and HTTPS_SERVER settings until someone manually changes one or the other. If they don't do that, the $config_secure and the secure method in $url->link ineffectual.

This leads to many insecure opencart installations (there are tons) across the web, as people end up setting the HTTP_SERVER to http:// as well, and 2.3 did things worse by forcing the secure flag $url->link for admin pages, even when the site doesn't have SSL available.

I think this is remedied by making the installation default to secure settings, and only unlockable by explicit action by the site administrator (rather than a user with administrative privileges). This would be through adding a FORCE_SSL global set to true in the config files, but which the admin could temporarily turn off in the event they needed to.

I would rather see us introduce blatant warnings to the user if they are insistent on running it over HTTP - adding a warning in the admin interface like it does if the install directory is still present, and even injecting messages into catalog pages (regardless of whether they're on a page we "think" should be secure).

As far as multi-store goes... Again, my thought mirrors this: SSL shouldn't be the option of the store admin, rather the site admin. That's what I'm hoping we move beyond.

Basically, make the site usable without HTTPS

**Comment by ADDCreative** [comment]
None of which requires the link protocol to be based to the request, https://github.com/copona/copona/commit/ac98570fedf86d9ec3ab7de5e3c059907e21092e Which is dangerous.

**Comment by arnisjuraga** [comment]
@ADDCreative Hi, can you check the latest commit, do you have any comments on that, please?

**Comment by ADDCreative** [comment]
@arnisjuraga Looks good.

The URL generation still needs fixing. I'm not sure how you would want it to work. Are dropping the ability to have just the checkout and account pages HTTPS? Do you want a multi-stores HTTPS setting per domain? Do you what to keep the configuration setting in the database? If HTTPS setting per domain then it will need to be via the database.

**Comment by arnisjuraga** [comment]
- ok, thanks for the feedback!
- yes, if HTTPS is enabled, then it's forced to be used everywhere.
- for Multistore - no idea! :)

 I have no positive experience with Opencart multi-stores, and I believe. Current URL and SEO generation needs huge improvements. So, as multi-stores is not a priority at the moment (until, there be someone, who would love to request that).

HTTPS usage requires SSL certificate, which can be implemented by SNI and so on. When anyone will have a problem with the multi-stores - I will make the changes, so that every store can have it's own url obviously.
